lib/models/room.js

```diff
@@ -56,18 +72,24 @@ function synthesizeReceipt(userId, event, receiptType) {
  * when the call to <code>sendEvent</code> was made. If "<b>end</b>", pending messages
  * will always appear at the end of the timeline (multiple pending messages will be sorted
  * chronologically). Default: "chronological".
+ * @param {boolean} [opts.timelineSupport = false] Set to true to enable improved
+ * timeline support.
+ *
  * @prop {string} roomId The ID of this room.
  * @prop {string} name The human-readable display name for this room.
- * @prop {Array<MatrixEvent>} timeline The ordered list of message events for
- * this room, with the oldest event at index 0.
+ * @prop {Array<MatrixEvent>} timeline The live event timeline for this room,
+ * with the oldest event at index 0. Present for backwards compatibility -
+ * prefer getLiveTimeline().getEvents().
```

Comment:
Sadly, it doesn't quite provide exact backwards-compat.

The problem is as follows, with an existing client which doesn't know about timelines:

* client does a sync which returns event A
* client goes to sleep for a while; meanwhile events B and C happen
* client does another sync, which has a gap since the previous event, and contains event D. The live timeline is therefore reset and contains only event D.
* client tries to back-paginate to see what happened before event D; it gets events A, B, and C back from the server.
* B and C are preprended to the live timeline, which is fine, but the room already knows about A, so the old timeline is joined to the new
* From the client's point of view (given it knows nothing about multiple timelines), it can't see event A (or anything which happened before it).

The solution I went with was just to make the room forget all existing events/timelines when it got a gappy sync if `timelineSupport` was unset. I guess an alternative would be to make `timeline` a property whose getter builds an array from the live timeline and any previous timelines which have been joined up. But I'd be worried about people assuming `room.timeline` was a cheap operation.